### PR TITLE
Writing tests for Month and Calendar Objects and their helper functions

### PR DESCRIPTION
**Pull request by andrayantelo**

### CODE REVIEW

**Comment by johntyree** [comment]
What am I reviewing here?

**Comment by andrayantelo** [comment]
I haven't gotten started on this one yet.

File: tests/tests.html [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  <!-- include font awesome -->
+  <link rel="stylesheet" href="//maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css">
+
+  <!-- Material Design -->
+  <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
+
+  <!-- Loading spinner -->
+  <script src="https://cdn.jsdelivr.net/spinjs/2.3.2/spin.min.js"></script>
+  
+  <script src="https://www.gstatic.com/firebasejs/3.6.5/firebase.js"></script>
```
Comment (johntyree) [inline tests/tests.html]: You don't want to be running tests against your production database.

Either use localstorage or figure out how to use firebase's testing facilities.

File: tests/tests.js [hunk 2]
```diff
@@ -0,0 +1,259 @@
+// Tests for checkit app
+
+// the test method gets passed a string as the first argument (naming the
+// test), and a function as the second (which will run the code for this test)
+// div #qunit-fixture contains an extract of the markup from our app, enough to 
+// write useful tests against. By putting it in this div, we don't have to worry
+// about DOM changes from one test affecting other tests because qunit will 
+// automatically reset the markup after each test.
+
+QUnit.test( "a basic test example", function( assert ) {
```
Comment (johntyree) [inline tests/tests.js]: Probably don't need this one :)

File: tests/tests.js [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    assert.equal(this.testmonth.firstDayIndex, 3, "Checking index of the first of the month");
+    
+    assert.deepEqual(this.testmonth.dayIndex, {}, "Initialized the day Index object");
+    assert.equal(this.testmonth.monthId, "20171", "Initialized the month Id");
+});
+
+
+// Testing Calendar Helper functions
+QUnit.module( "Calendar helper functions tests", {
+});
```
Comment (johntyree) [inline tests/tests.js:44]: Is this supposed to be around these tests?

File: tests/tests.js [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+QUnit.test("generateUniqueId test", function(assert) {
+    assert.expect(1);
+    var uniqueId = generateUniqueId();
+    var uniqueId2 = generateUniqueId();
+    
+    assert.notEqual(uniqueId, uniqueId2);
+    // Anything else I should be testing in here?
```
Comment (johntyree) [inline tests/tests.js]: Seems ok

File: tests/tests.js [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+    assert.equal(this.calendar.state.endDateString, "20170320", "Check calendar endDateString added a month");
+    assert.equal(this.calendar.monthObjects.length, 2, "Check month was added to monthObjects array");
+    assert.equal(typeof(this.calendar.addMonth()), "object", "addMonth returns an object");
+});
+
+QUnit.test("generateMonthObjects calendar method test", function(assert) {
+    
+    assert.expect(5);
+    
+    var monthObjects = this.calendar.generateMonthObjects(moment("20170101", "YYYYMMDD"), moment("20170214", "YYYYMMDD"));
```
Comment (johntyree) [inline tests/tests.js]: You also want to test corner cases and errors. What if the start and end dates are the same? What if they are one day apart? What if the end date is earlier than start date? Does the function return the "right" answer in all these cases?

**Review by johntyree — CHANGES_REQUESTED** [review]

File: tests/tests.js (hunk 3 above)
Comment (andrayantelo) [reply tests/tests.js:44]: No.

Comment (andrayantelo) [reply tests/tests.js:44]: It's weird.

File: tests/tests.html (hunk 1 above)
Comment (andrayantelo) [reply tests/tests.html]: If I don't include this script, then I can't use auth either. Should I not use auth either?

File: tests/template.html [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+        <td>
+          <div class="nill"></div>
+        </td>
+    </tr>
+    <tr class="week">
+      <td>
+        <div class="nill"></div>
+      </td>
+      <td>
+        <div class="nill"></div>
```
Comment (johntyree) [inline tests/template.html]: DIdn't this get changed to `.nil` at some point?

File: tests/tests.js [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+
+QUnit.test("emptyCalendarState test", function(assert) {
+    assert.expect(6);
+    var params = {startDate: "2017-02-14" , endDate: "2017-02-20" , calendarTitle: ""};
+    var state = emptyCalendarState(params);
+    
+    assert.equal(state.title, "", "Checking state title");
+    assert.equal(state.startDateString, "20170214", "Checking startDateString");
+    assert.equal(state.endDateString, "20170220", "Checking endDateString");
+    // What kind of assertion for uniqueId?
```
Comment (johntyree) [inline tests/tests.js]: non-empty is fine

File: tests/tests.js [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+    assert.deepEqual(state.checkedDays, {});
+    
+    //assert that an exception is thrown with invalid dates.
+    var params = {startDate: "2017-02-14", endDate: "2017-01-01", calendarTitle: "hello"};
+    // I don't understand the qunit api for assert.raises (or assert.throws)
```
Comment (johntyree) [inline tests/tests.js]: What's your question?

File: tests/tests.js [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+    assert.expect(7);
+    
+    var monthObjects = this.calendar.generateMonthObjects(moment("20170101", "YYYYMMDD"), moment("20170214", "YYYYMMDD"));
+
+    assert.ok(Array.isArray(monthObjects));
+    assert.equal(monthObjects[0].dateString, "20170101");
+    assert.equal(monthObjects[1].dateString, "20170201");
+    assert.equal(monthObjects[1].lastActiveDay, 14);
+    assert.equal(monthObjects.length, 2);
+    // Don't understand why this assertion passes
```
Comment (johntyree) [inline tests/tests.js]: These assertions check to make sure that your function fails as expected. The assertion succeeds because the function failed in the way you described.

**Review by johntyree — CHANGES_REQUESTED** [review]
If you aren't explicitly testing the database you probably don't want to be initializing and connecting to firebase every time you run your tests.

You should be able to test your code against a mock database and test the database separately.

File: tests/tests.js (hunk 8 above)
Comment (andrayantelo) [reply tests/tests.js]: How does it work? Although, I followed along with the example they had when I wrote this.

File: tests/template.html (hunk 6 above)
Comment (andrayantelo) [reply tests/template.html]: Fixed.

File: docs/checkit.js [hunk 10]
```diff
@@ -732,7 +742,8 @@ CheckIt.prototype.findCurrentDay = function() {
 
 $(document).ready(function() {
     
-    checkit = new CheckIt();
+    checkit = new CheckIt('localStorage');
```
Comment (johntyree) [inline docs/checkit.js]: Won't this break everything? I thought we only wanted to use localStorage in tests.

Comment (johntyree) [reply docs/checkit.js]: Oh I see the problem, this is in `checkit.js`. You probably want this module to have all the code you need to *define* a Checkit, but not actually create one. The code to create the object can either be in a `main.js` or `app.js` or whatever the cool kids are calling their entry point these days, or even just in a `<script>` tag on the main page, that way you can have `mode = 'firebase'` there and `mode = 'localStorage'` in your test html.

File: docs/checkit.js [hunk 11]
```diff
@@ -773,10 +788,17 @@ var generateUniqueId = function() {
 };
 
 var emptyCalendarState = function(params) {
+    var startDate = moment(params.startDate, "YYYY-MM-DD");
+    var endDate = moment(params.endDate, "YYYY-MM-DD");
+    // Should an error be thrown if the calendar title parameter is an empty string?
```
Comment (johntyree) [inline docs/checkit.js]: I probably would.

**Review by johntyree — CHANGES_REQUESTED** [review]

File: docs/checkit.js [hunk 12]
```diff
@@ -84,7 +84,12 @@ function CheckIt() {
     
     // We have to connect to Firebase before we can access it.
     // WHEN PAGE LOADS
-    this.initFirebase();
+    if (this.mode === 'firebase') {
+        this.initFirebase();
+    }
+    else {
```
Comment (johntyree) [inline docs/checkit.js]: I'd probably explicitly check for `'localStorage`' and throw otherwise. That protects you against typos like `new Checkit('firebsae');`.
```
switch (this.mode) {
  case 'firebase': this.initFirebase(); break;
  case 'localStorage': this.initLocalStorage(); break;
  default: throw new Error("Unknown storage mode: '" + this.mode + "'");
}
```

**Review by johntyree — CHANGES_REQUESTED** [review]

File: docs/checkit.js [hunk 13]
```diff
@@ -82,13 +82,11 @@ function CheckIt(mode) {
...
     
-    // We have to connect to Firebase before we can access it.
-    // WHEN PAGE LOADS
-    if (this.mode === 'firebase') {
-        this.initFirebase();
-    }
-    else {
-        this.initLocalStorage();
+    // Initialize storage
+    switch (this.mode) {
```
Comment (johntyree) [inline docs/checkit.js]: Did you test it?

**Review by johntyree — CHANGES_REQUESTED** [review]

File: docs/template.html [hunk 14]
```diff
@@ -0,0 +1,162 @@
+'<div class="header">\
```
Comment (johntyree) [inline docs/template.html]: You can remove this file now.

File: docs/checkit.js [hunk 15]
```diff
@@ -82,9 +83,12 @@ function CheckIt() {
     this.$signInButton.click(this.signIn.bind(this));
    
     
-    // We have to connect to Firebase before we can access it.
-    // WHEN PAGE LOADS
-    this.initFirebase();
+    // Initialize storage
+    switch (this.mode) {
```
Comment (johntyree) [inline docs/checkit.js:87]: Where is the test for this?

**Review by johntyree — CHANGES_REQUESTED** [review]

**Review by johntyree — APPROVED** [review]
